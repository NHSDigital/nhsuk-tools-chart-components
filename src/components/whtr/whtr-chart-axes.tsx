import React, { JSX } from 'react';
import { WhtrResultChart } from '@models/whtr-chart.model';
import { RiskClassifications } from '@custom-types/whtr.types';
import './styles/whtr-chart-axes.module.scss';

type WhtrtGraphAxisMarkerProps = {
  markerText: string;
  offsetPercent: number;
};

type WhtrtGraphAxesProps = {
  model: WhtrResultChart;
};

const markerAxisLabelClassname = (offsetPercent: number): string =>
  offsetPercent > 0
    ? 'nhsuk-whtr-axis-marker-label'
    : 'nhsuk-whtr-axis-start-marker-label';

const markerAxisClassname = (offsetPercent: number): string =>
  offsetPercent > 0 ? 'nhsuk-whtr-axis-marker' : 'nhsuk-whtr-start-axis-marker';

function AxisMarker({
  markerText,
  offsetPercent,
}: Readonly<WhtrtGraphAxisMarkerProps>): JSX.Element {
  return (
    <div
      className={`${markerAxisClassname(offsetPercent)}`}
      style={{ left: `${offsetPercent}%` }}>
      <div
        className={`nhsuk-font-size-16 nhsuk-u-secondary-text-color ${markerAxisLabelClassname(
          offsetPercent
        )}`}>
        {markerText}
      </div>
    </div>
  );
}

export function WhtrChartAxes({
  model,
}: Readonly<WhtrtGraphAxesProps>): JSX.Element {
  return (
    <>
      <AxisMarker
        markerText={model.bounds.lowerBound.toString()}
        offsetPercent={model.calculatePosition(model.bounds.lowerBound)}
      />
      <AxisMarker
        markerText={model.bounds.upperBounds[
          RiskClassifications.Low
        ].toString()}
        offsetPercent={model.classificationPosition(RiskClassifications.Low)}
      />
      <AxisMarker
        markerText={model.bounds.upperBounds[
          RiskClassifications.Increased
        ].toString()}
        offsetPercent={model.classificationPosition(
          RiskClassifications.Increased
        )}
      />
    </>
  );
}
