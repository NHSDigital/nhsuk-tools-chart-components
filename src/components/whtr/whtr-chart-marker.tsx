import React, { JSX } from 'react';
import { WhtrResultChart } from '@models/whtr-chart.model';
import { ArrowMarkerIcon } from '@components/markers/arrow-marker-icon';
import './styles/whtr-chart-marker.module.scss';

const markerAlighTextClassname = (offset: number): string => {
  if (offset <= 20) {
    return 'nhsuk-whtr-chart-label-align-left';
  }
  if (offset <= 80) {
    return 'nhsuk-whtr-chart-label-align-center';
  }
  return 'nhsuk-whtr-chart-label-align-right';
};

function WhtrChartMarkerText({
  ratio,
  offset,
  markerText,
}: Readonly<{
  ratio: number;
  offset: number;
  markerText: string;
}>): JSX.Element {
  return (
    <div className="nhsuk-whtr-chart-label-marker">
      <div className={`nhsuk-font-size-16 ${markerAlighTextClassname(offset)}`}>
        {markerText}
        <span className="nhsuk-whtr-chart-ratio-marker">{ratio}</span>
      </div>
    </div>
  );
}
export function WhtrChartMarker({
  ratio,
  model,
  markerText,
}: Readonly<{
  ratio: number;
  model: WhtrResultChart;
  markerText: string;
}>): JSX.Element {
  const offset = model.clampedOffsetPercentage(ratio);

  const roundingPrecision = 1;
  const multiplier = 10 ** roundingPrecision;
  const roundedOffset = Math.round(offset * multiplier) / multiplier;

  return (
    <div>
      <div
        className="nhsuk-whtr-chart-arrow-marker"
        style={{ left: `${roundedOffset}%` }}>
        <WhtrChartMarkerText
          ratio={ratio}
          markerText={markerText}
          offset={roundedOffset}
        />
        <ArrowMarkerIcon />
      </div>
      <div
        className="nhsuk-whtr-chart-axis-marker"
        style={{ left: `${roundedOffset}%` }}
      />
    </div>
  );
}
