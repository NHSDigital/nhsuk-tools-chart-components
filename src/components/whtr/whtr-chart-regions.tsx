import React, { JSX } from 'react';
import { WhtrResultChart } from '@models/whtr-chart.model';
import {
  RiskClassification,
  RiskClassifications,
} from '@custom-types/whtr.types';
import './styles/whtr-chart-regions.module.scss';

function RegionForClassification({
  classification,
  model,
}: Readonly<{
  classification: RiskClassification;
  model: WhtrResultChart;
}>): JSX.Element {
  const width = model.classificationPosition(classification);
  return (
    <div
      className={`nhsuk-whtr-classification-${classification}`}
      style={{ width: `${width}%` }}
    />
  );
}

export function WhtrChartRegions({
  model,
}: Readonly<{
  model: WhtrResultChart;
}>): JSX.Element {
  return (
    <>
      <RegionForClassification
        classification={RiskClassifications.Low}
        model={model}
      />
      <RegionForClassification
        classification={RiskClassifications.Increased}
        model={model}
      />
      <RegionForClassification
        classification={RiskClassifications.High}
        model={model}
      />
    </>
  );
}
