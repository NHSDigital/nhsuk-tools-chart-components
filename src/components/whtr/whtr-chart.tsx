import { WhtrResultChart } from '@models/whtr-chart.model';
import React, { JSX } from 'react';
import { RiskClassificationBounds } from '@custom-types/whtr.types';
import { WhtrChartMarker } from './whtr-chart-marker';
import { WhtrChartRegions } from './whtr-chart-regions';
import { WhtrChartAxes } from './whtr-chart-axes';
import './styles/whtr-chart.module.scss';

const CHART_WIDTH = 100;

/**
 * The properties for @WhtrChart component
 *
 * @param {RiskClassificationBounds} classificationBounds -  The Waist to height classification ranges
 * @param {number} ratio - The waist to height ratio value
 * @param {string} markerText - Marker text for the ratio value
 */
export type WhtrChartProps = {
  classificationBounds: RiskClassificationBounds;
  ratio: number;
  markerText: string;
};

/**
 * A component to display the waist to height chart
 *
 * @typedef {object} WhtrChartProps
 * @property {RiskClassificationBounds} classificationBounds -  The waist to height classification ranges
 * @property {number} ratio - The waist to height ratio value
 * @property {string} markerText - Marker text for the ratio value
 *
 * @param {WhtrChartProps} props
 * @returns {JSX.Element}
 */
export function WhtrChart({
  classificationBounds,
  ratio,
  markerText,
}: Readonly<WhtrChartProps>): JSX.Element {
  const chartModel = new WhtrResultChart(classificationBounds, CHART_WIDTH);

  return (
    <div
      className={`nhsuk-whtr-graph nhsuk-u-margin-bottom-9 `}
      aria-hidden="true">
      <WhtrChartAxes model={chartModel} />
      <WhtrChartMarker
        ratio={ratio}
        markerText={markerText}
        model={chartModel}
      />
      <WhtrChartRegions model={chartModel} />
    </div>
  );
}
