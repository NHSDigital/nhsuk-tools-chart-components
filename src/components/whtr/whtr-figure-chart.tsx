import React, { JSX } from 'react';
import { ChartLegend } from '@components/chart-legend';
import {
  RiskClassificationBounds,
  WhtrLegendKeys,
} from '@custom-types/whtr.types';
import { WhtrChart } from './whtr-chart';
import './styles/whtr-chart-card.module.scss';

/**
 * The properties for @WhtrChartFigureProps component
 *
 * @param {RiskClassificationBounds} classificationBounds -  The waist to height classification ranges
 * @param {number} ratio - The waist to height ratio value
 * @param {string} ariaLabel - The aria label for the chart figure
 * @param {string} ratioMarkerText - The marker text for the waist to height ratio value
 * @param {WhtrLegendKeys} legendKeys - The chart legend items
 */
export type WhtrChartFigureProps = {
  classificationBounds: RiskClassificationBounds;
  ratio: number;
  ariaLabel: string;
  ratioMarkerText: string;
  legendKeys: WhtrLegendKeys;
};

/**
 * A component to display the waist to height chart figure
 *
 * @typedef {object} WhtrChartFigureProps
 * @property {RiskClassificationBounds} classificationBounds -  The waist to height classification ranges
 * @property {number} ratio - The waist to height ratio value
 * @property {string} ratioMarkerText - The marker text for the waist to height ratio value
 * @property {string} ariaLabel - The aria label for the chart figure
 * @property {WhtrLegendKeys} legendKeys - The chart legend items
 *
 * @param {WhtrChartFigureProps} props
 * @returns {JSX.Element}
 */
export function WhtrChartFigure({
  classificationBounds,
  ratio,
  ratioMarkerText,
  ariaLabel,
  legendKeys,
}: Readonly<WhtrChartFigureProps>): JSX.Element {
  const whtrLegendKeys = Object.entries(legendKeys).map(
    ([classification, label]) => ({
      text: label,
      className: `nhsuk-whtr-legend-key-classification-${classification}`,
      wrapperClassName: 'nhsuk-whtr-legend-marker',
    })
  );

  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <WhtrChart
        classificationBounds={classificationBounds}
        ratio={ratio}
        markerText={ratioMarkerText}
      />
      <ChartLegend legendKeys={whtrLegendKeys} borderWrapper={false} />
    </figure>
  );
}
