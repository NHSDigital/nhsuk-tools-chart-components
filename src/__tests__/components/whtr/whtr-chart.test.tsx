import { render, screen } from '@testing-library/react';
import React, { JSX } from 'react';
import '@testing-library/jest-dom';
import {
  mockedWtRatioValue,
  mockedWtrClassificationBounds,
} from '@tests/__helpers__/mocks/whtr-results';
import { WhtrChart } from '@components/whtr/whtr-chart';
import { WeightedBounds } from '@custom-types/threshold.types';
import { RiskClassification } from '@custom-types/whtr.types';

const mockedMarkerText = 'The ratio:';

function ResultWhtrResultChartTestApp({
  classificationBounds = mockedWtrClassificationBounds,
  ratio = mockedWtRatioValue,
  markerText = mockedMarkerText,
}: Readonly<{
  classificationBounds?: WeightedBounds<RiskClassification>;
  ratio?: number;
  markerText?: string;
}>): JSX.Element {
  return (
    <WhtrChart
      classificationBounds={classificationBounds}
      ratio={ratio}
      markerText={markerText}
    />
  );
}

describe('The Waist to height result chart component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<ResultWhtrResultChartTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders the markerText', () => {
    const { container } = render(
      <ResultWhtrResultChartTestApp markerText={mockedMarkerText} />
    );

    const labelMarkerEl = container?.querySelector(
      '.nhsuk-whtr-chart-label-marker'
    );
    expect(labelMarkerEl).not.toBeNull();
    expect(labelMarkerEl?.innerHTML).toContain(mockedMarkerText);

    expect(screen.getByText(mockedMarkerText).classList[1]).toContain(
      'nhsuk-whtr-chart-label-align-right'
    );
  });

  it.each`
    ratioValue | expectedAxisPosition
    ${1}       | ${98}
    ${0.7}     | ${98}
    ${0.68}    | ${98}
    ${0.65}    | ${98}
    ${0.64}    | ${95.2}
    ${0.55}    | ${57}
    ${0.5}     | ${38}
    ${0.49}    | ${34.2}
    ${0.45}    | ${19}
    ${0.41}    | ${3.8}
    ${0.4}     | ${0}
    ${0.39}    | ${0}
    ${0.35}    | ${0}
    ${0.1}     | ${0}
  `(
    'expect the chart market axis to be positioned relative to ratio value with $expectedAxisPosition',
    async ({
      ratioValue,
      expectedAxisPosition,
    }: {
      ratioValue: number;
      expectedAxisPosition: number;
    }) => {
      const { container } = render(
        <ResultWhtrResultChartTestApp ratio={ratioValue} />
      );

      const axisMarkerEl = container?.querySelector(
        '.nhsuk-whtr-chart-axis-marker'
      ) as HTMLElement;
      expect(axisMarkerEl).not.toBeNull();
      expect(axisMarkerEl.style.left).toBe(`${expectedAxisPosition}%`);

      const arrowMarkerEl = container?.querySelector(
        '.nhsuk-whtr-chart-arrow-marker'
      ) as HTMLElement;
      expect(arrowMarkerEl).not.toBeNull();
      expect(arrowMarkerEl.style.left).toBe(`${expectedAxisPosition}%`);
    }
  );

  it('Renders the chart regions', () => {
    const { container } = render(<ResultWhtrResultChartTestApp />);

    const lowRiskRegionEl = container?.querySelector(
      '.nhsuk-whtr-classification-low'
    ) as HTMLElement;
    expect(lowRiskRegionEl).not.toBeNull();
    expect(lowRiskRegionEl.style.width).toBe('38%');

    const increasedRiskRegionEl = container?.querySelector(
      '.nhsuk-whtr-classification-increased'
    ) as HTMLElement;
    expect(increasedRiskRegionEl).not.toBeNull();
    expect(increasedRiskRegionEl.style.width).toBe('76%');

    const highRiskRegionEl = container?.querySelector(
      '.nhsuk-whtr-classification-high'
    ) as HTMLElement;
    expect(highRiskRegionEl).not.toBeNull();
    expect(highRiskRegionEl.style.width).toBe('100%');
  });

  it('Renders the chart axis ', () => {
    const { container } = render(<ResultWhtrResultChartTestApp />);

    const axisElements = container?.querySelectorAll('.nhsuk-whtr-axis-marker');

    expect(axisElements.length).toBe(2);
    const lowRiskAxisEl = axisElements[0] as HTMLElement;
    const increasedRiskAxisEl = axisElements[1] as HTMLElement;

    expect(lowRiskAxisEl.style.left).toBe('38%');
    expect(increasedRiskAxisEl.style.left).toBe('76%');

    const startAxisLabelElement = container?.querySelector(
      '.nhsuk-whtr-axis-start-marker-label'
    );
    expect(startAxisLabelElement!.innerHTML).toBe('0.4');

    const axisLabelElements = container?.querySelectorAll(
      '.nhsuk-whtr-axis-marker-label'
    );

    expect(axisLabelElements.length).toBe(2);
    const lowRiskAxisLabelEl = axisLabelElements[0];
    const increasedRiskAxisLabelEl = axisLabelElements[1];
    expect(lowRiskAxisLabelEl.innerHTML).toBe('0.5');

    expect(increasedRiskAxisLabelEl.innerHTML).toBe('0.6');
  });

  it.each`
    ratioValue
    ${0.6}
    ${0.61}
    ${0.56}
    ${0.45}
    ${0.4}
  `(
    'expect the chart to contain the ratio label value as $ratioValue',
    async ({ ratioValue }: { ratioValue: number }) => {
      const { container } = render(
        <ResultWhtrResultChartTestApp ratio={ratioValue} />
      );

      const ratioMarkerEl = container?.querySelector(
        '.nhsuk-whtr-chart-ratio-marker'
      );
      expect(ratioMarkerEl).not.toBeNull();
      expect(ratioMarkerEl?.innerHTML).toContain(ratioValue.toString());
    }
  );
});
