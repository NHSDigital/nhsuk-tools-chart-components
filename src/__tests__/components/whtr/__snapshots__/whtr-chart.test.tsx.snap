// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Waist to height result chart component Renders with the expected content 1`] = `
<div>
  <div
    aria-hidden="true"
    class="nhsuk-whtr-graph nhsuk-u-margin-bottom-9 "
  >
    <div
      class="nhsuk-whtr-start-axis-marker"
      style="left: 0%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-whtr-axis-start-marker-label"
      >
        0.4
      </div>
    </div>
    <div
      class="nhsuk-whtr-axis-marker"
      style="left: 38%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-whtr-axis-marker-label"
      >
        0.5
      </div>
    </div>
    <div
      class="nhsuk-whtr-axis-marker"
      style="left: 76%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-whtr-axis-marker-label"
      >
        0.6
      </div>
    </div>
    <div>
      <div
        class="nhsuk-whtr-chart-arrow-marker"
        style="left: 95.2%;"
      >
        <div
          class="nhsuk-whtr-chart-label-marker"
        >
          <div
            class="nhsuk-font-size-16 nhsuk-whtr-chart-label-align-right"
          >
            The ratio:
            <span
              class="nhsuk-whtr-chart-ratio-marker"
            >
              0.64
            </span>
          </div>
        </div>
        <svg
          fill="none"
          viewBox="0 0 12 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 9L0.803849 -9.78799e-07L11.1962 -7.02746e-08L6 9Z"
            fill="black"
          />
        </svg>
      </div>
      <div
        class="nhsuk-whtr-chart-axis-marker"
        style="left: 95.2%;"
      />
    </div>
    <div
      class="nhsuk-whtr-classification-low"
      style="width: 38%;"
    />
    <div
      class="nhsuk-whtr-classification-increased"
      style="width: 76%;"
    />
    <div
      class="nhsuk-whtr-classification-high"
      style="width: 100%;"
    />
  </div>
</div>
`;
